lib/honeybadger/notice.ex

```diff
@@ -1,43 +1,76 @@
 defmodule Honeybadger.Notice do
   alias Honeybadger.Utils
 
+  @typep error :: %{class: atom | iodata, message: iodata, tags: list, backtrace: list}
+
+  @type noticeable :: Exception.t | Map.t | String.t | atom
+
+  @typep notifier :: %{name: String.t, url: String.t, version: String.t}
+
+  @typep server :: %{environment_name: atom, hostname: String.t, project_root: Path.t}
+
+  @type t :: %__MODULE__{
+    notifier: notifier,
+    server: server,
+    error: error,
+    request: Map.t
+  }
+
+  @enforce_keys [:notifier, :server, :error, :request]
+
   defstruct [:notifier, :server, :error, :request]
 
+  @doc """
+  Create a new `Honeybadger.Notice` struct for various error types.
+
+  ## Example
+
+      iex> Honeybadger.Notice.new("oops", %{}, []).error
+      %{backtrace: [], class: "RuntimeError", message: "oops", tags: []}
 
-  def new(error, metadata \\ %{}, backtrace)
+      iex> Honeybadger.Notice.new(:badarg, %{}, []).error
+      %{backtrace: [], class: "ArgumentError", message: "argument error", tags: []}
+
+      iex> Honeybadger.Notice.new(%RuntimeError{message: "oops"}, %{}, []).error
+      %{backtrace: [], class: "RuntimeError", message: "oops", tags: []}
+  """
+  @spec new(noticeable, Map.t, list) :: t
+  def new(error, metadata, backtrace)
+
+  def new(message, metadata, backtrace) when is_binary(message) do
+    new(%RuntimeError{message: message}, metadata, backtrace)
+  end
 
   def new(%{class: class, message: message}, metadata, backtrace) do
-    error = %{
-      class: class,
-      message: message,
-      tags: Map.get(metadata, :tags, []),
-      backtrace: backtrace
-    }
-    create(error, metadata)
+    %{class: class, message: message, backtrace: backtrace}
+    |> create(metadata)
   end
 
   def new(exception, metadata, backtrace) do
     exception = Exception.normalize(:error, exception)
-    exception_mod = exception.__struct__
+
+    %{__struct__: exception_mod} = exception
+
     error = %{
       class: Utils.module_to_string(exception_mod),
       message: exception_mod.message(exception),
-      tags: Map.get(metadata, :tags, []),
       backtrace: backtrace
     }
+
     create(error, metadata)
   end
 
   defp create(error, metadata) do
+    error = Map.put(error, :tags, Map.get(metadata, :tags, []))
     context = Map.get(metadata, :context, %{})
-    request = Map.get(metadata, :plug_env, %{})
-              |> Map.put(:context, context)
-
-    %__MODULE__{error: error,
-                request: request,
-                notifier: notifier(),
-                server: server()}
-     |> filter(Honeybadger.get_env(:notice_filter))
+    request =
+      metadata
+      |> Map.get(:plug_env, %{})
+      |> Map.put(:context, context)
+
+    __MODULE__
+    |> struct!(error: error, request: request, notifier: notifier(), server: server())
```

Comment:
Using `%Notice{error: ... }` may be more readable than using `struct!`

Reply:
Using `%Notice{}` is definitely more readable. I was aiming to leverage `@enforce_keys`, but that is checked at compile time for literal constructors anyhow. This has been changed (with an `alias __MODULE__` at the top to avoid writing `%__MODULE__{}`